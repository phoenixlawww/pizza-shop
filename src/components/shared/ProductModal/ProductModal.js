import React, { useContext, useImperativeHandle, useState } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { find, isArray } from 'lodash'
import ReactModal from 'react-modal'
import { useForm } from 'react-hook-form'
import { yupResolver } from '@hookform/resolvers/yup'
import * as yup from 'yup'

import ShopContext from 'contexts/ShopContext'

import Button, { BUTTON_VARIENTS } from 'components/shared/Button'

import './ProductModal.scss'

const ProductModal = React.forwardRef((props, ref) => {
  const { defaultShow, product, productOptions } = props

  const { dispatch } = useContext(ShopContext)

  const [isShow, setIsShow] = useState(defaultShow)

  const show = () => {
    setIsShow(true)
  }
  const hide = () => {
    setIsShow(false)
  }

  useImperativeHandle(ref, () => ({
    show,
    hide
  }))

  const schema = yup.object().shape(
    (() => {
      if (!productOptions) {
        return {}
      }

      return productOptions.reduce((objectSchema, option) => {
        const { category, allowMultiple } = option

        objectSchema[category] = allowMultiple
          ? yup.array().nullable(true)
          : yup.string().required()

        return objectSchema
      }, {})
    })()
  )

  const defaultValues = (() => {
    if (!productOptions) {
      return {}
    }

    return productOptions.reduce((objectSchema, option) => {
      const { category, allowMultiple } = option

      objectSchema[category] = allowMultiple ? [] : null

      return objectSchema
    }, {})
  })()

  const { formState, register, handleSubmit, reset } = useForm({
    mode: 'onChange',
    defaultValues,
    resolver: yupResolver(schema)
  })

  const onSubmit = (data) => {
    const selectedOptions = []

    Object.keys(data).forEach((category) => {
      const selections = isArray(data[category])
        ? data[category]
        : [data[category]]

      const subOptions =
        find(productOptions, {
          category
        })?.subOptions || []

      selections.forEach((selection) => {
        if (selection) {
          selectedOptions.push(find(subOptions, { code: selection }))
        }
      })
    })

    dispatch({
      type: 'ADD_PRODUCT_TO_CART',
      product: {
        ...product,
        options: selectedOptions,
        quantity: 1
      }
    })

    setIsShow(false)
    reset()
  }

  if (!product || !productOptions) {
    return null
  }

  return (
    <ReactModal
      isOpen={isShow}
      // appElement={document.getElementById('root')}
      ariaHideApp={false}
      style={{
        overlay: {
          backgroundColor: 'rgba(0,0,0,0.2)'
        },
        content: {
          height: 'fit-content',
          maxWidth: 600,
          margin: '0 auto'
        }
      }}
    >
      <form onSubmit={handleSubmit(onSubmit)}>
        <div className="s-product-modal">
          <div className="product-options">
            {productOptions.map((option, index) => {
              const { category, allowMultiple, subOptions } = option

              return (
                <div className="product-option" key={category}>
                  <h5>{category}</h5>
                  <div
                    className={classnames('product-sub-option-list', {
                      multiple: allowMultiple,
                      single: !allowMultiple
                    })}
                  >
                    {subOptions.map((subOption, index) => {
                      const { code, name } = subOption

                      if (allowMultiple) {
                        return (
                          <div className="product-sub-option" key={index}>
                            <input
                              id={code}
                              type="checkbox"
                              // name={category}
                              value={code}
                              {...register(category)}
                            />
                            <label htmlFor={code}>{name}</label>
                          </div>
                        )
                      } else {
                        return (
                          <div className="product-sub-option" key={index}>
                            <input
                              id={code}
                              type="radio"
                              // name={category}
                              value={code}
                              {...register(category)}
                            />
                            <label htmlFor={code}>{name}</label>
                          </div>
                        )
                      }
                    })}
                  </div>
                </div>
              )
            })}
          </div>
          <div className="action-buttons">
            <Button
              className="action-btn"
              label={'Add To Basket'}
              varient={BUTTON_VARIENTS.PRIMARY}
              disabled={!formState.isValid}
              type="submit"
            />
            <Button
              className="action-btn"
              label={'Cancel'}
              varient={BUTTON_VARIENTS.SECONDARY}
              onClick={(e) => {
                setIsShow(false)
                reset()
              }}
            />
          </div>
        </div>
      </form>
    </ReactModal>
  )
})

ProductModal.displayName = 'ProductModal'

ProductModal.propTypes = {
  defaultShow: PropTypes.bool,
  product: PropTypes.object,
  productOptions: PropTypes.array
}
ProductModal.defaultProps = {
  defaultShow: false
}

export default ProductModal
