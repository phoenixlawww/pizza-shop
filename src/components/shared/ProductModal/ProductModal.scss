@import "~styles/tools";

.s-product-modal {
  display: flex;
  flex-direction: column;
  align-items: stretch;
  justify-content: space-between;
  height: 100%;

  > .product-options {
    h5 {
      margin-top: 0;
      margin-bottom: 10px;
    }

    .product-option {
      + .product-option {
        margin-top: 10px;
      }
    }
  }

  .product-sub-option-list {
    &.multiple {
      columns: 3;
    }

    &.single {
    }

    .product-sub-option {
      padding-right: 10px;

      + .product-sub-option {
        margin-top: 10px;
      }
    }

    input[type="radio"],
    input[type="checkbox"] {
      display: none;

      + label {
        position: relative;
        display: flex;
        flex-direction: row;
        align-items: flex-start;
        justify-content: flex-start;
        line-height: 18px;
        cursor: pointer;

        &::before {
          display: inline-block;
          flex-shrink: 0;

          width: 1rem;
          height: 1rem;
          margin-right: 10px;
          overflow: hidden;
          font-size: 16px;
          line-height: 1;

          color: $black;
          text-align: center;
          content: "";
          border: 1px solid $black;
          border-radius: 3px;
        }
      }
    }

    input[type="radio"] {
      + label {
        &::before {
          border-radius: 50%;
        }
      }

      &:checked {
        + label {
          &::before {
            box-shadow: inset 0 0 0 4px $black;
          }
        }
      }
    }

    input[type="checkbox"] {
      + label {
        &::before {
          border-radius: 3px;
        }
      }

      &:checked {
        + label {
          &::before {
            font-weight: 600;
            content: "\2713";
          }
        }
      }
    }
  }

  > .action-buttons {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    align-items: stretch;
    justify-content: space-between;
    margin-top: 20px;

    > .action-btn {
      min-width: 200px;
      margin-right: 10px;
      margin-bottom: 10px;

      // +.action-btn {
      //   margin-top: 10px;
      // }
    }
  }
}
