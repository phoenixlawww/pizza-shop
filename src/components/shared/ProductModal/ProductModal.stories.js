import { useRef } from 'react'

import Button from 'components/shared/Button'
import ProductModal from './ProductModal'

import ShopContext from 'contexts/ShopContext'

import pizzaList from 'mock-data/products/pizza.json'
import pizzaOptions from 'mock-data/products/pizza/options.json'

export default {
  component: ProductModal,
  title: 'Shared/ProductModal',
  args: {
    defaultShow: true,
    product: pizzaList.products[0],
    productOptions: pizzaOptions.options
  },
  decorators: [
    (Story) => (
      <ShopContext.Provider
        value={{
          dispatch: (action) => {
            alert(JSON.stringify(action))
          }
        }}
      >
        <Story />
      </ShopContext.Provider>
    )
  ]
}

const Template = (args) => {
  const modalRef = useRef()

  return (
    <>
      <Button
        label="Open"
        onClick={() => {
          modalRef.current.show()
        }}
      />
      <ProductModal {...args} ref={modalRef} />
    </>
  )
}

export const Normal = Template.bind({})
Normal.args = {}
