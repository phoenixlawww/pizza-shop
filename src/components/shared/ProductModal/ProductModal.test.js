import { render } from '@testing-library/react'

import ProductModal from './ProductModal'

import pizzaList from 'mock-data/products/pizza.json'
import pizzaOptions from 'mock-data/products/pizza/options.json'

const testProps = {
  defaultShow: true,
  product: pizzaList.products[0],
  productOptions: pizzaOptions.options
}

test('renders correctly', () => {
  const { baseElement } = render(<ProductModal {...testProps} />)

  expect(baseElement).toMatchSnapshot()
})
