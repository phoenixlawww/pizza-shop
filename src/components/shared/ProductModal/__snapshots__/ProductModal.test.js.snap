// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<body
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div />
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay"
      style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(0, 0, 0, 0.2);"
    >
      <div
        aria-modal="true"
        class="ReactModal__Content"
        role="dialog"
        style="position: absolute; top: 40px; left: 40px; right: 40px; bottom: 40px; border: 1px solid #ccc; background: rgb(255, 255, 255); overflow: auto; border-radius: 4px; outline: none; padding: 20px; max-width: 600px; margin: 0px auto;"
        tabindex="-1"
      >
        <form>
          <div
            class="s-product-modal"
          >
            <div
              class="product-options"
            >
              <div
                class="product-option"
              >
                <h5>
                  topping
                </h5>
                <div
                  class="product-sub-option-list multiple"
                >
                  <div
                    class="product-sub-option"
                  >
                    <input
                      id="topping-1"
                      name="topping"
                      type="checkbox"
                      value="topping-1"
                    />
                    <label
                      for="topping-1"
                    >
                      Pizza Topping #1
                    </label>
                  </div>
                  <div
                    class="product-sub-option"
                  >
                    <input
                      id="topping-2"
                      name="topping"
                      type="checkbox"
                      value="topping-2"
                    />
                    <label
                      for="topping-2"
                    >
                      Pizza Topping #2
                    </label>
                  </div>
                  <div
                    class="product-sub-option"
                  >
                    <input
                      id="topping-3"
                      name="topping"
                      type="checkbox"
                      value="topping-3"
                    />
                    <label
                      for="topping-3"
                    >
                      Pizza Topping #3
                    </label>
                  </div>
                  <div
                    class="product-sub-option"
                  >
                    <input
                      id="topping-4"
                      name="topping"
                      type="checkbox"
                      value="topping-4"
                    />
                    <label
                      for="topping-4"
                    >
                      Pizza Topping #4
                    </label>
                  </div>
                  <div
                    class="product-sub-option"
                  >
                    <input
                      id="topping-5"
                      name="topping"
                      type="checkbox"
                      value="topping-5"
                    />
                    <label
                      for="topping-5"
                    >
                      Pizza Topping #5
                    </label>
                  </div>
                  <div
                    class="product-sub-option"
                  >
                    <input
                      id="topping-6"
                      name="topping"
                      type="checkbox"
                      value="topping-6"
                    />
                    <label
                      for="topping-6"
                    >
                      Pizza Topping #6
                    </label>
                  </div>
                  <div
                    class="product-sub-option"
                  >
                    <input
                      id="topping-7"
                      name="topping"
                      type="checkbox"
                      value="topping-7"
                    />
                    <label
                      for="topping-7"
                    >
                      Pizza Topping #7
                    </label>
                  </div>
                  <div
                    class="product-sub-option"
                  >
                    <input
                      id="topping-8"
                      name="topping"
                      type="checkbox"
                      value="topping-8"
                    />
                    <label
                      for="topping-8"
                    >
                      Pizza Topping #8
                    </label>
                  </div>
                  <div
                    class="product-sub-option"
                  >
                    <input
                      id="topping-9"
                      name="topping"
                      type="checkbox"
                      value="topping-9"
                    />
                    <label
                      for="topping-9"
                    >
                      Pizza Topping #9
                    </label>
                  </div>
                </div>
              </div>
              <div
                class="product-option"
              >
                <h5>
                  size
                </h5>
                <div
                  class="product-sub-option-list single"
                >
                  <div
                    class="product-sub-option"
                  >
                    <input
                      id="size-s"
                      name="size"
                      type="radio"
                      value="size-s"
                    />
                    <label
                      for="size-s"
                    >
                      Pizza Size #1
                    </label>
                  </div>
                  <div
                    class="product-sub-option"
                  >
                    <input
                      id="size-m"
                      name="size"
                      type="radio"
                      value="size-m"
                    />
                    <label
                      for="size-m"
                    >
                      Pizza Size #2
                    </label>
                  </div>
                  <div
                    class="product-sub-option"
                  >
                    <input
                      id="size-l"
                      name="size"
                      type="radio"
                      value="size-l"
                    />
                    <label
                      for="size-l"
                    >
                      Pizza Size #3
                    </label>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="action-buttons"
            >
              <button
                class="s-button action-btn s-button-primary"
                disabled=""
                type="submit"
              >
                Add To Basket
              </button>
              <button
                class="s-button action-btn s-button-secondary"
              >
                Cancel
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</body>
`;
